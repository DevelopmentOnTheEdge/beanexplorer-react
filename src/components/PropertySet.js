import React from 'react';
import PropTypes from 'prop-types';
import Property from './Property';
import classNames from 'classnames';
import JsonPointer from 'json-pointer';
import {shouldPropertyUpdate} from "./utils";

class PropertySet extends React.Component {

  shouldComponentUpdate(nextProps) {
    return shouldPropertyUpdate(this.props, nextProps) || this.props.values !== nextProps.values;
  }

  static getName(name) {
    if (name) {
      return <h5 className='property-group__title'>{name}</h5>
    } else {
      return null
    }
  }

  createGroup(curGroup, curGroupId, curGroupName, curGroupClasses) {
    return (
      <div
        className={classNames(
          'property-group',
          curGroupClasses || 'property-group__top-line col-12'
        )}
        key={curGroupId}
        ref={curGroupId}
      >
        <div className="property-group__top-line-row row"/>
        {PropertySet.getName(curGroupName)}
        <div className={classNames('property-group__row', this.props.rowClass)}>
          {curGroup}
        </div>
      </div>
    );
  }

  render() {
    let curGroup = [];
    let curGroupName = null, curGroupId = null, curGroupClasses = null;
    let fields = [];

    const finishGroup = () => {
      if (curGroup.length > 0) {
        if (curGroupId) {
          fields.push(this.createGroup(curGroup, curGroupId, curGroupName, curGroupClasses));
        } else {
          Array.prototype.push.apply(fields, curGroup);
        }
      }
      curGroup = [];
    };

    for (const path of this.props.bean.order) {
      const meta = this.props.bean.meta[path];

      const newGroupId = meta.groupId;
      const newGroupName = meta.groupName;
      const newGroupClasses = meta.groupClasses;

      if (newGroupId !== curGroupId) {
        finishGroup();
        curGroupName = newGroupName;
        curGroupClasses = newGroupClasses;
        curGroupId = newGroupId;
      }

      const field = (
        <Property key={path} path={path} {...this.props} value={this.getValue(path)}/>
      );

      curGroup.push(field);
    }
    finishGroup();

    return (
      <div className={classNames('property-set', this.props.rowClass)}>
        {fields}
      </div>
    );
  }

  getValue(path) {
    const values = this.props.values || this.props.bean.values;
    return JsonPointer.get(values, path)
  }

}

PropertySet.defaultProps = {
  rowClass: 'form-row'
};

PropertySet.propTypes = {
  bean: PropTypes.object.isRequired,
  values: PropTypes.object,
  onChange: PropTypes.func,
  inline: PropTypes.bool,
  horizontal: PropTypes.bool,
  bsSize: PropTypes.string,
  localization: PropTypes.object,
  rowClass: PropTypes.string
};

export default PropertySet;
