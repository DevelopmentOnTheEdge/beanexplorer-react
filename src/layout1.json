{
  "values": {
    "name": "",
    "subject": ""
  },
  "meta": {
    "/name": {
      "placeholder": "Name"
    },
    "/subject": {
      "placeholder": "Subject"
    }
  },
  "order": [
    "/name",
    "/subject"
  ]
}