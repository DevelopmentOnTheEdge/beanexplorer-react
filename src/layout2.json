{
  "values": {
    "name2": "value",
    "subject2": "",
    "labelField": "Text label",
    "hiddenInput": ""
  },
  "meta": {
    "/name2": {
      "displayName": "Name",
      "inputSize": 10
    },
    "/subject2": {
      "displayName": "Remember me",
      "type": "Boolean"
    },
    "/labelField": {
      "labelField": true
    },
    "/hiddenInput": {
      "placeholder": "hidden",
      "hidden": true
    }
  },
  "order": [
    "/name2",
    "/subject2",
    "/labelField",
    "/hiddenInput"
  ]
}