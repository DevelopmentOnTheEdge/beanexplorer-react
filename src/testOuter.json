{
  "values": {
    "textInput": "test",
    "checkBox": true,
    "labelField": "Pellentesque habitant morbi tristique senectus...",
    "textInput2": "test",
    "checkBox2": false,
    "textInput3": "test"
  },
  "meta": {
    "/textInput": {
      "displayName": "Name",
      "groupName": "Group 1",
      "groupId": 1
    },
    "/checkBox": {
      "displayName": "CheckBox",
      "type": "Boolean",
      "groupId": 1
    },
    "/labelField": {
      "labelField": true,
      "groupId": 1
    },
    "/textInput2": {
      "displayName": "Name",
      "groupName": "Group 2 - With Statuses",
      "groupId": 2,
      "message": "Error here",
      "status": "error"
    },
    "/checkBox2": {
      "displayName": "CheckBox",
      "type": "Boolean",
      "groupId": 2,
      "status": "success"
    },
    "/textInput3": {
      "displayName": "Name",
      "groupName": "Group 3 - CanBeNull",
      "groupId": 3,
      "canBeNull": true
    }
  },
  "order": [
    "/textInput",
    "/checkBox",
    "/labelField",
    "/textInput2",
    "/checkBox2",
    "/textInput3"
  ]
}