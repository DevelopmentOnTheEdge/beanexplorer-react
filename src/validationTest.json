{
  "values": {
    "short": 32767,
    "integer": 2147483647,
    "long": 9007199254740991,
    "double": 100.1,
    "doubleWithStep": 100.5,
    "integerWithRangeAndStep": "",
    "textInput": "qwerty7890"
  },
  "meta": {
    "/short": {
      "displayName": "Short",
      "type": "Short",
      "groupName": "Numbers",
      "groupId": 1
    },
    "/integer": {
      "displayName": "Integer",
      "type": "Integer",
      "groupId": 1
    },
    "/long": {
      "displayName": "Long",
      "type": "Long",
      "message": "use Number.MAX_SAFE_INTEGER and Number.MIN_SAFE_INTEGER",
      "groupId": 1
    },
    "/double": {
      "displayName": "Double",
      "type": "Double",
      "message": "have problems in firefox",
      "groupId": 1
    },
    "/doubleWithStep": {
      "displayName": "Double with custom step",
      "type": "Double",
      "validationRules":[{"type":"step","attr":0.5},{"type":"range","attr":{"min":0,"max":1000}}],
      "message": "Use also the 'range' for correct work in chrome, firefox, etc",
      "groupId": 1
    },
    "/integerWithRangeAndStep": {
      "displayName": "Number input with range and integer step",
      "validationRules":[{"type":"step","attr":10},{"type":"range","attr":{"min":0,"max":100}}],
      "groupId": 1,
      "canBeNull": true
    },
    "/textInput": {
      "displayName": "Text input with columnSize",
      "columnSize": 10,
      "groupName": "Text",
      "groupId": 2
    }
  },
  "order": [
    "/short",
    "/integer",
    "/long",
    "/double",
    "/doubleWithStep",
    "/integerWithRangeAndStep",
    "/textInput"
  ]
}