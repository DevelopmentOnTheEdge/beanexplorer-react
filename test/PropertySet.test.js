import React from 'react';
import PropertySet from '../src/components/PropertySet';
import renderer from 'react-test-renderer';
import {shallow, mount, render} from 'enzyme';
import bean from '../src/testJson.json';
import testOuter from '../src/testOuter.json';
import validationTest from '../src/validationTest.json';
import layout1 from '../src/layout1.json';
import layout2 from '../src/layout2.json';

//https://github.com/YouCanBookMe/react-datetime/issues/384
jest.mock('react-dom', () => ({
  findDOMNode: () => {},
}));

test('renders without crashing bean ', () => {

  mount(<PropertySet bean={bean}/>);

  mount(<PropertySet bean={bean} bsSize="sm"/>);
  mount(<PropertySet bean={bean} bsSize="lg"/>);

});

test('snapshot bean', () => {
  const component = renderer.create(
    <PropertySet bean={bean} />
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('snapshot bean horizontal', () => {
  const component = renderer.create(
    <PropertySet bean={bean} horizontal/>
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('snapshot bean with readOnly', () => {
  for(let item in bean.meta) {
    bean.meta[item]['readOnly'] = true;
  }

  const component = renderer.create(
    <PropertySet bean={bean} />
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('renders without crashing empty value', () => {
  for(let item in bean.value) {
    bean.value[item] = "";
  }

  mount(<PropertySet bean={bean}/>);
});

test('test outer', () => {
  const component = renderer.create(
    <PropertySet bean={testOuter} />
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('validationTest', () => {
  const component = renderer.create(
    <PropertySet bean={validationTest} />
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('inline layout1', () => {
  const component = renderer.create(
    <PropertySet bean={layout1} inline rowClass="" />
  );
  expect(component.toJSON()).toMatchSnapshot();

  //renders without crashing
  mount(<PropertySet bean={layout1} inline rowClass="" bsSize="sm"/>);
  mount(<PropertySet bean={layout1} inline rowClass="" bsSize="lg"/>);
});

test('inline layout2', () => {
  const component = renderer.create(
    <PropertySet bean={layout2} inline rowClass="" />
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('simple property set', () => {
	const simpleBean = {
    "values": { "number": "" },
    "meta":   { "/number": {} },
    "order":  [ "/number" ]
	};

	const component = renderer.create(
    <PropertySet bean={simpleBean} />
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('empty property set', () => {
	const emptyBean = {
		 "values": {},
		 "meta": {},
		 "order": []
	};

	const component = renderer.create(
    <PropertySet bean={emptyBean} />
  );
  expect(component.toJSON()).toMatchSnapshot();
});
