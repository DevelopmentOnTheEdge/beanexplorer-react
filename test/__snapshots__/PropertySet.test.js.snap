// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`empty property set 1`] = `
<div
  className="property-set"
/>
`;

exports[`simple property set 1`] = `
<div
  className="property-set"
>
  <div
    className="form-group property "
  >
    <div
      className="input"
    >
      <label
        className={undefined}
        htmlFor="numberField"
      >
        numberField
      </label>
      <div
        className="controls"
      >
        <input
          className="form-control"
          id="numberField"
          onChange={[Function]}
          placeholder={undefined}
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;
